jni/src/knn_extension/faiss/utils/BitSet.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <algorithm>
+#include "BitSet.h"
+
+FixedBitSet::FixedBitSet(const int* int_array, const int length){
+    assert(int_array && "int_array should not be null");
+    const int* maxValue = std::max_element(int_array, int_array + length);
+    this->numBits = (*maxValue >> 6) + 1; // div by 64
+    this->bitmap = new uint64_t[this->numBits]();
+    for(int i = 0 ; i < length ; i ++) {
+        int value = int_array[i];
+        int bitsetArrayIndex = value >> 6;
+        this->bitmap[bitsetArrayIndex] |= 1ULL << (value & 63); // Equivalent of 1ULL << (value % 64)
+    }
+}
+
+idx_t FixedBitSet::next_set_bit(idx_t index) const {
+    idx_t i = index >> 6; // div by 64
+    uint64_t word = this->bitmap[i] >> (index & 63); // Equivalent of bitmap[i] >> (index % 64)
+
+    if (word != 0) {
+      return index + __builtin_ctzll(word);
```

Comment:
Could you add a comment around usage of this? What does word represent? It looks like its the 64-bit int that is like the "bit page" shifted over offset bits.

Reply:
Take a look at the comment in Heap.h file and let me know if you still need some comment here to understand the code.

Reply:
Not sure which comment in Heap.h you are referring

Reply:
I mean
```
 * Here a block is 64 bit. However, for simplicity let's assume its size is 8 bits.
 * Then, if have an array of 3, 7, and 10, it will be represented in bitmap as follow.
 *            [0]      [1]
 * bitmap: 10001000 00000100
 *
 * for next_set_bit call with 4
 * 1. it looks for bitmap[0]
 * 2. bitmap[0] >> 4
 * 3. count trailing zero of the result from step 2 which is 3
 * 4. return 4(current index) + 3(result from step 3)
 ```

Reply:
Thats not in Heap.h, but makes sense

Reply:
Sorry. BitSet.h :)